### ISSUE TITLE
pass request_id as `x-request-id` header to action/remote schema

### ISSUE THREAD

**Issue by albttx** [hasura/graphql-engine#6654]
I'd like to be able to trace a full request from my frontend -> hasura -> api when i call an hasura action.

There is the `x-request-id` header for that.

If i set it on my frontend query, as it's specified the same `x-request-id` will be in hasura logs, but hasura don't send this header to my API.

But i receive a 2 tracing headers: `X-Hasura-Traceid` and  `X-Hasura-Spanid`. unfortunately those id aren't logged in hasura log  ouput.

On a action called, hasura forward all the http headers to the api

- if i sent from my frontend a http header `x-foobar-id: 42`, my API will receive it.
- if i sent from my frontend a http header `x-request-id: 42`, my API will NOT receive it.

Am i missing something ?!

Tested on `v1.3.3` and on `v2.0.0-alpha.4`

**Comment by tirumaraiselvan** [comment]
First, you will have to check "Forward Client Headers to webhook" option in your action. Then, do you get other headers forwarded from the client? For example, what if you set `random-header: abcd` in the client.

**Comment by albttx** [comment]
@tirumaraiselvan Yes and yes.

- The `Forward Client Headers to webhook` is set to `true`

as i show i my example, any other "random header" are passed successfully.

>>>
-   if i sent from my frontend a http header x-foobar-id: 42, my API will receive it.
-    if i sent from my frontend a http header x-request-id: 42, my API will NOT receive it.
>>>

**Comment by tirumaraiselvan** [comment]
Thanks, we will investigate.

**Comment by tirumaraiselvan** [comment]
@albttx I am not able to reproduce this with an action handler running locally. 

This is my metadata (you can import it yourself):

```
{
  "version": 2,
  "tables": [],
  "actions": [
    {
      "name": "actionName",
      "definition": {
        "handler": "http://172.17.0.1:3000/createUser",
        "output_type": "SampleOutput",
        "forward_client_headers": true,
        "arguments": [
          {
            "name": "arg1",
            "type": "SampleInput!"
          }
        ],
        "type": "mutation",
        "kind": "synchronous"
      }
    }
  ],
  "custom_types": {
    "input_objects": [
      {
        "name": "SampleInput",
        "fields": [
          {
            "name": "username",
            "type": "String!"
          },
          {
            "name": "password",
            "type": "String!"
          }
        ]
      }
    ],
    "objects": [
      {
        "name": "SampleOutput",
        "fields": [
          {
            "name": "userId",
            "type": "Int!"
          }
        ]
      }
    ]
  }
}
```

and this is my webhook code:

```
app.post('/createUser', (req, res) => {
    console.log(req.body);
    console.log(req.headers);
    res.json({userId: 1});
});
```

Response (you can see **x-request-id** here):

```
{
  session_variables: { 'x-hasura-role': 'admin' },
  input: { arg1: { username: 'klsdjflak', password: 'Adfadf' } },
  action: { name: 'actionName' }
}
{
  host: '172.17.0.1:3000',
  'accept-encoding': 'gzip',
  'content-length': '140',
  'x-hasura-traceid': '10092408740106224238',
  'x-hasura-spanid': '10258280524722883924',
  'content-type': 'application/json',
  'user-agent': 'hasura-graphql-engine/v1.3.3',
  'x-forwarded-user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36',
  'sec-fetch-mode': 'cors',
  'sec-fetch-site': 'same-origin',
  'x-request-id': 'adfadfa',
  'x-forwarded-host': 'localhost:8080',
  cookie: '_ga=GA1.1.521871140.1571028955; intercom-id-rucirpb3=fe2333da-c879-4eaa-9113-15894e235eec; _lfa=eyJYYnAxb2FFMnByZzhFZFZqIjoiTEYxLjEuOTg5NzE3MWZlZWMzY2E2Mi4xNTkwNTgwNjM2NDU4In0%3D; hubspotutk=442884a356ff9485879b7e7bb1dadee9; __hstc=181257784.442884a356ff9485879b7e7bb1dadee9.1590580637148.1593414561426.1593434752398.7; ajs_user_id=null; ajs_group_id=null; driftt_aid=b6eb371a-31f9-40fc-ba9d-4632a9e50e3d; DFTT_END_USER_PREV_BOOTSTRAPPED=true'
}
```

Since `x-request-id` is kind of a special header recognized by various API gateways, maybe it is being stripped at a different location.

**Comment by albttx** [comment]
Hello @tirumaraiselvan, i destroy and re-pull my hasura docker image while i was trying to create a separate repository to send you de replicated issue but then it worked ! :tada:
I really don't understand what happened.. 

I keep you in touch in case the issue is back, but i can close for now

sorry for disturbing you, and thanks a lot for this amazing product :pray:

**Comment by albttx** [comment]
Hello @tirumaraiselvan , the issue came back... and 

I managed to reproduce the error and i created a small example to reproduce easily 

https://github.com/albttx/hasura-issue-6654

Could you take a look ?

**Comment by SidharthBihary** [comment]
Hey @albttx , I tried passing 'x-request-id' header to actions with the action-handler being set up on localhost, Heroku and glitch. The header value is set to `'x-request-id': 'abcd'` on client. 

For every query, the hasura log has `'x-request-id': 'abcd'` in it's request_headers. So, hasura makes the http-request including the 'x-request-id' header.

Now for the action handler:

- When the action handler is on localhost, the action-handler receives the header with the value set on client, i.e. `'x-request-id': 'abcd'`.
- However, when the action handler is on Heroku, the action-handler receives the header but with a random value(which is different from the value set on client), `'x-request-id': '1d286958-b6fe-46dc-a42b-3e0234cf4e8a'`.
- And for the action handler on glitch, it does not receive the 'x-request-id'. 

So, hasura does include the `'x-request-id'` header in the http-request it makes, but it depends on the platform gateway to reject or forward it.

**Comment by albttx** [comment]
Hello @SidharthBihary thanks for answering.

Exactly, if you send the send header `x-request-id: abcd` from the client to hasura, hasura will logs the `request_id` as `abcd` and will sent it to the API.

But if you don't sent a `x-request-id`, for example if you use the GraphiQL playground to call your action, Hasura will generate a `x-request-id`, and will logs it, but the header won't be sent to the API.

> So, hasura does include the 'x-request-id' header in the http-request it makes, but it depends on the platform gateway to reject or forward it.

Do you said that is a normal behavior ? 

Why does this depends on the platform ? Hasura action should always share the `x-request-id` to the api...

ps: i tests all cases on a separate repository https://github.com/albttx/hasura-issue-6654 with details in the `README.md`.

**Comment by tirumaraiselvan** [comment]
@albttx I think the confusion was that hasura logs `request_id` (not x-request-id). Indeed, we should pass `request_id` into `x-request-id` header. We will add this ability ASAP. 

Thanks for bringing it to our notice.

**Comment by ecthiender** [comment]
Fixed via a83194239ebfa26c76d20f0d90e5513225f497df
